PR: https://github.com/odoo/odoo/pull/160187

From: 0da3259034ff3b2b79f417df53969fe017207195
From: Mohammad Abdulmoneim (abdu)
Date: 2026-04-15 10:56:45

Breaking data model changes scores: del:2 + add:0, change matches:
-    is_pickup_location = fields.Boolean()
-    pickup_location_data = fields.Json()

Total Changes: 124

[IMP] *: use new pickup location selector in backend

*: delivery, sale, stock_delivery, website_sale

This commit moves all the code related to pickup points feature from
`delivery` module to `website_sale_stock`. This is because there's no
need to have the feature in `delivery`, since all delivery methods that
currently support this feature depend on `website_sale_stock`, namely:
Sendcloud, FedEx and Click&Collect.

This commit introduces several changes and improvements to pickup point
delivery feature.

** Behavior after this commit:
* The json field is removed from the sales order.
* When the user confirms an order with a pickup delivery, a new address
  is created with the point data (if there's no existing address with
  the same data).
* A new json field is added to `res.partner` to capture the info needed
  by the shipping provider later when processing the shipment.
* Pickup address can be changed from sales/delivery order using the new
  location selector (introduced in #156382) which is currently
  used in eCommerce.

Task-3645144

Part-of: odoo/odoo#160187
Related: odoo/enterprise#59935
Related: odoo/upgrade#6315
Signed-off-by: Valeriya Chuprina (vchu) <vchu@odoo.com>

================================= pseudo patch: =================================

--- a/addons/delivery/models/res_partner.py
+++ b/addons/delivery/models/res_partner.py
@@ -1,7 +1,6 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 from odoo import fields, models
-from odoo.fields import Domain
 
 
 class ResPartner(models.Model):
@@ -13,7 +12,3 @@ class ResPartner(models.Model):
         comodel_name="delivery.carrier",
         company_dependent=True,
     )
-    is_pickup_location = fields.Boolean()  # Whether it is a pickup point address.
-
-    def _get_delivery_address_domain(self):
-        return super()._get_delivery_address_domain() & Domain("is_pickup_location", "=", False)

--- a/addons/delivery/models/sale_order.py
+++ b/addons/delivery/models/sale_order.py
@@ -1,7 +1,5 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-import json
-
 from odoo import _, api, fields, models
 from odoo.exceptions import UserError
 
@@ -9,7 +7,6 @@ from odoo.exceptions import UserError
 class SaleOrder(models.Model):
     _inherit = "sale.order"
 
-    pickup_location_data = fields.Json()
     carrier_id = fields.Many2one(
         string="Delivery Method",
         help="Fill this field if you plan to invoice the shipping based on picking.",
@@ -24,13 +21,6 @@ class SaleOrder(models.Model):
     )
     shipping_weight = fields.Float(compute="_compute_shipping_weight", store=True, readonly=False)
 
-    def _compute_partner_shipping_id(self):
-        """Override to reset the delivery address when a pickup location was selected."""
-        super()._compute_partner_shipping_id()
-        for order in self:
-            if order.partner_shipping_id.is_pickup_location:
-                order.partner_shipping_id = order.partner_id
-
     @api.depends("order_line")
     def _compute_is_service_products(self):
         for so in self:
@@ -95,58 +85,6 @@ class SaleOrder(models.Model):
             order._create_delivery_line(carrier, amount)
         return True
 
-    def _set_pickup_location(self, pickup_location_data):
-        """Set the pickup location on the current order.
-
-        Note: self.ensure_one()
-
-        :param str pickup_location_data: The JSON-formatted pickup location address.
-        :return: None
-        """
-        self.ensure_one()
-        use_locations_fname = f"{self.carrier_id.delivery_type}_use_locations"
-        if hasattr(self.carrier_id, use_locations_fname):
-            use_location = getattr(self.carrier_id, use_locations_fname)
-            if use_location and pickup_location_data:
-                pickup_location = json.loads(pickup_location_data)
-            else:
-                pickup_location = None
-            self.pickup_location_data = pickup_location
-
-    def _get_pickup_locations(self, zip_code=None, country=None, **kwargs):
-        """Return the pickup locations of the delivery method close to a given zip code.
-
-        Use provided `zip_code` and `country` or the order's delivery address to determine the zip
-        code and the country to use.
-
-        Note: self.ensure_one()
-
-        :param int zip_code: The zip code to look up to, optional.
-        :param res.country country: The country to look up to, required if `zip_code` is provided.
-        :return: The close pickup locations data.
-        :rtype: dict
-        """
-        self.ensure_one()
-        if country:
-            partner_address = self.env["res.partner"].new({
-                "active": False,
-                "country_id": country.id,
-                "zip": zip_code,
-            })
-        else:
-            partner_address = self.partner_shipping_id
-        try:
-            error = {"error": _("No pick-up points are available for this delivery address.")}
-            function_name = f"_{self.carrier_id.delivery_type}_get_close_locations"
-            if not hasattr(self.carrier_id, function_name):
-                return error
-            pickup_locations = getattr(self.carrier_id, function_name)(partner_address, **kwargs)
-            if not pickup_locations:
-                return error
-            return {"pickup_locations": pickup_locations}
-        except UserError as e:
-            return {"error": str(e)}
-
     def action_open_delivery_wizard(self):
         view_id = self.env.ref("delivery.choose_delivery_carrier_view_form").id
         if self.env.context.get("carrier_recompute"):
@@ -175,63 +113,6 @@ class SaleOrder(models.Model):
             },
         }
 
-    def _action_confirm(self):
-        for order in self:
-            order_location = order.pickup_location_data
-
-            if not order_location:
-                continue
-
-            # Retrieve all the data : name, street, city, state, zip, country.
-            name = order_location.get("name") or order.partner_shipping_id.name
-            street = order_location["street"]
-            city = order_location["city"]
-            zip_code = order_location["zip_code"]
-            country_code = order_location["country_code"]
-            country = order.env["res.country"].search([("code", "=", country_code)]).id
-            state = None
-            if order_location.get("state") and country:
-                state = (
-                    order
-                    .env["res.country.state"]
-                    .search([("code", "=", order_location["state"]), ("country_id", "=", country)])
-                    .id
-                )
-            parent_id = order.partner_shipping_id.id
-            email = order.partner_shipping_id.email
-            phone = order.partner_shipping_id.phone
-
-            # Check if the current partner has a partner of type 'delivery' with the same address.
-            existing_partner = order.env["res.partner"].search(
-                [
-                    ("street", "=", street),
-                    ("city", "=", city),
-                    ("state_id", "=", state),
-                    ("country_id", "=", country),
-                    ("parent_id", "=", parent_id),
-                    ("type", "=", "delivery"),
-                ],
-                limit=1,
-            )
-
-            shipping_partner = existing_partner or order.env["res.partner"].create({
-                "parent_id": parent_id,
-                "type": "delivery",
-                "name": name,
-                "street": street,
-                "city": city,
-                "state_id": state,
-                "zip": zip_code,
-                "country_id": country,
-                "email": email,
-                "phone": phone,
-                "is_pickup_location": True,
-            })
-            order.with_context(update_delivery_shipping_partner=True).write({
-                "partner_shipping_id": shipping_partner
-            })
-        return super()._action_confirm()
-
     def _prepare_delivery_line_vals(self, carrier, price_unit):
         context = {}
         if self.partner_id:
